frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1417 @@
+		/// The total count of child keys in the migration, over all executions.
+		///
+		/// This only kept around for bookkeeping and debugging.
+		pub(crate) child_items: u32,
+
+		#[codec(skip)]
+		pub(crate) _ph: sp_std::marker::PhantomData<T>,
+	}
+
+	impl<T: Config> sp_std::fmt::Debug for MigrationTask<T> {
+		fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+			f.debug_struct("MigrationTask")
+				.field(
+					"top",
+					&self.current_top.as_ref().map(|d| sp_core::hexdisplay::HexDisplay::from(d)),
+				)
+				.field(
+					"child",
+					&self.current_child.as_ref().map(|d| sp_core::hexdisplay::HexDisplay::from(d)),
+				)
+				.field("prev_tick_child", &self.prev_tick_child)
+				.field("dyn_top_items", &self.dyn_top_items)
+				.field("dyn_child_items", &self.dyn_child_items)
+				.field("dyn_size", &self.dyn_size)
+				.field("size", &self.size)
+				.field("top_items", &self.top_items)
+				.field("child_items", &self.child_items)
+				.finish()
+		}
+	}
+
+	impl<T: Config> Default for MigrationTask<T> {
+		fn default() -> Self {
+			Self {
+				current_top: Some(Default::default()),
+				current_child: Default::default(),
+				dyn_child_items: Default::default(),
+				dyn_top_items: Default::default(),
+				dyn_size: Default::default(),
+				prev_tick_child: Default::default(),
+				_ph: Default::default(),
+				size: Default::default(),
+				top_items: Default::default(),
+				child_items: Default::default(),
+			}
+		}
+	}
+
+	impl<T: Config> MigrationTask<T> {
+		/// Return true if the task is finished.
+		pub(crate) fn finished(&self) -> bool {
+			self.current_top.is_none() && self.current_child.is_none()
+		}
+
+		/// Check if there's any work left, or if we have exhausted the limits already.
+		fn exhausted(&self, limits: MigrationLimits) -> bool {
+			self.current_top.is_none() ||
+				self.dyn_total_items() >= limits.item ||
+				self.dyn_size >= limits.size
+		}
+
+		/// get the total number of keys affected by the current task.
+		pub(crate) fn dyn_total_items(&self) -> u32 {
+			self.dyn_child_items.saturating_add(self.dyn_top_items)
+		}
+
+		/// Migrate keys until either of the given limits are exhausted, or if no more top keys
+		/// exist.
+		///
+		/// Note that this can return after the **first** migration tick that causes exhaustion,
+		/// specifically in the case of the `size` constrain. The reason for this is that before
+		/// reading a key, we simply cannot know how many bytes it is. In other words, this should
+		/// not be used in any environment where resources are strictly bounded (e.g. a parachain),
+		/// but it is acceptable otherwise (relay chain, offchain workers).
+		pub(crate) fn migrate_until_exhaustion(&mut self, limits: MigrationLimits) {
+			log!(debug, "running migrations on top of {:?} until {:?}", self, limits);
+
+			if limits.item.is_zero() || limits.size.is_zero() {
+				// handle this minor edge case, else we would call `migrate_tick` at least once.
```

Comment:
Can you not move the check in the loop above the `migrate_tick()` call?

Reply:
Can skip the edge case code then? (maybe I missed something)

Reply:
The problem with that is that `fn exhausted` expects the task to be initialized, as in `current_top` to be set to something other than `None`. If we first check exhaustion, it will mistakenly think that it is initially exhausted and just return `true`. And to fix that we need to add another edges case.

TLDR; we remove one edge case to add another one. No difference, let's stick to this :D